components/style/rule_tree/mod.rs

```diff
@@ -0,0 +1,491 @@
+
+        Some(current)
+    }
+
+    pub unsafe fn gc(&self) {
+        while let Some(weak) = self.pop_from_free_list() {
+            let needs_drop = {
+                let node = &*weak.ptr();
+                if node.refcount.load(Ordering::SeqCst) == 0 {
+                    node.remove_from_child_list();
+                    true
+                } else {
+                    false
+                }
+            };
+
+            debug!("GC'ing {:?}: {}", weak.ptr(), needs_drop);
+            if needs_drop {
+                let _ = Box::from_raw(weak.ptr());
+            }
+        }
+
+        debug_assert!(self.root.get().next_free.load(Ordering::SeqCst) == FREE_LIST_SENTINEL);
+    }
+}
+
+struct RuleNode {
+    /// The root node. Only the root has no root pointer, for obvious reasons.
+    root: Option<StrongRuleNode>,
+
+    /// The parent rule node. Only the root has no parent.
+    parent: Option<StrongRuleNode>,
+
+    /// The actual style rule that matched. Only the root has no declarations.
+    rule: Option<Arc<StyleRule>>,
+
+    /// The importance of the declarations relevant in the style rule,
+    /// meaningless in the root node.
+    importance: Importance,
+
+    children: RuleChildrenList,
+    refcount: AtomicUsize,
+    next: AtomicPtr<RuleNode>,
+    prev: AtomicPtr<RuleNode>,
+
+    /// The next item in the rule tree free list, that starts on the root node.
+    next_free: AtomicPtr<RuleNode>,
+}
+
+unsafe impl Sync for RuleTree {}
+unsafe impl Send for RuleTree {}
+
+impl RuleNode {
+    fn new(root: StrongRuleNode,
+           parent: StrongRuleNode,
+           style_rule: Arc<StyleRule>,
+           importance: Importance) -> Self {
+        debug_assert!(root.get().parent.is_none());
+        RuleNode {
+            root: Some(root),
+            parent: Some(parent),
+            rule: Some(style_rule),
+            importance: importance,
+            children: RuleChildrenList::new(),
+            refcount: AtomicUsize::new(1),
+            next: AtomicPtr::new(ptr::null_mut()),
+            prev: AtomicPtr::new(ptr::null_mut()),
+            next_free: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+
+    fn root() -> Self {
+        RuleNode {
+            root: None,
+            parent: None,
+            rule: None,
+            importance: Importance::Normal,
+            children: RuleChildrenList::new(),
+            refcount: AtomicUsize::new(1),
+            next: AtomicPtr::new(ptr::null_mut()),
+            prev: AtomicPtr::new(ptr::null_mut()),
+            next_free: AtomicPtr::new(FREE_LIST_SENTINEL),
+        }
+    }
+
+    /// Remove this rule node from the child list.
+    ///
+    /// This method doesn't use proper synchronization, and it's expected to be
+    /// called in a single-threaded fashion, thus the unsafety.
+    ///
+    /// This is expected to be called before freeing the node from the free
+    /// list.
+    unsafe fn remove_from_child_list(&self) {
```

Comment:
In debug mode, could you set a `sweeping_in_progress` flag here and have all of the list methods check that flag and panic if it's set?

I know it's paranoid, but if it ever trips in the test suite we'll be *really* glad that we added that flag.